File: SenLib/Sen1/FileFixes/scripts_scena_dat_us_a0006_dat.cs
using HyoutaUtils;
using System.Collections.Generic;
using System.IO;

namespace SenLib.Sen1.FileFixes {
	public class scripts_scena_dat_us_a0006_dat : FileMod {
		public string GetDescription() {
			return "Fix double spaces in debug map.";
		}

		public IEnumerable<FileModResult> TryApply(FileStorage storage) {
			var s = storage.TryGetDuplicate(new HyoutaUtils.Checksum.SHA1(0x73fd289006017eabul, 0x8816636d998d21aaul, 0x8fb38d68u));
			if (s == null) {
				return null;
			}
			MemoryStream bin = s.CopyToMemoryAndDispose();

			var patcher = new SenScriptPatcher(bin);
			patcher.RemovePartialCommand(0x3a67, 0x6f, 0x3ac4, 1);
			patcher.RemovePartialCommand(0x3e5f, 0x5f, 0x3e8c, 1);

			return new FileModResult[] { new FileModResult("data/scripts/scena/dat_us/a0006.dat", bin) };
		}

		public IEnumerable<FileModResult> TryRevert(FileStorage storage) {
			var s = storage.TryGetDuplicate(new HyoutaUtils.Checksum.SHA1(0x73fd289006017eabul, 0x8816636d998d21aaul, 0x8fb38d68u));
			if (s == null) {
				return null;
			}
			return new FileModResult[] { new FileModResult("data/scripts/scena/dat_us/a0006.dat", s) };
		}
	}
}

File: native/sen1/file_fixes/scripts_scena_dat_us_a0006_dat.cpp
#include <string_view>
#include <vector>

#include "sen/file_getter.h"
#include "p3a/pack.h"
#include "p3a/structs.h"
#include "sen/sen_script_patcher.h"
#include "util/hash/sha1.h"

namespace SenLib::Sen1::FileFixes::scripts_scena_dat_us_a0006_dat {
std::string_view GetDescription() {
    return "Fix formatting issues in debug map.";
}

bool TryApply(const SenPatcher::GetCheckedFileCallback& getCheckedFile,
              std::vector<SenPatcher::P3APackFile>& result) {
    try {
        auto file = getCheckedFile(
            "data/scripts/scena/dat_us/a0006.dat",
            28887,
            SenPatcher::SHA1FromHexString("73fd289006017eab8816636d998d21aa8fb38d68"));
        if (!file) {
            return false;
        }

        auto& bin = file->Data;

        SenScriptPatcher patcher(bin);
        patcher.RemovePartialCommand(0x3a67, 0x6f, 0x3ac4, 1);
        patcher.RemovePartialCommand(0x3e5f, 0x5f, 0x3e8c, 1);

        result.emplace_back(std::move(bin), file->Filename, SenPatcher::P3ACompressionType::LZ4);

        return true;
    } catch (...) {
        return false;
    }
}
} // namespace SenLib::Sen1::FileFixes::scripts_scena_dat_us_a0006_dat
